contrib/upgradability-patterns.asciidoc

```diff
@@ -0,0 +1,367 @@
+   * @param record sha3 key
+   * @param value new value is bytes
+   */
+  function setBytesValue(bytes32 record, bytes value) public onlyOwner {
+    BytesValues[record] = value;
+  }
+
+  /**
+   * @dev delete the value inside BytesValues for the given key 
+   * @param record sha3 key
+   */
+  function deleteBytesValue(bytes32 record) public onlyOwner {
+    delete BytesValues[record];
+  }
+
+  /**
+   * @dev change the value inside Bytes32Values for the given key 
+   * @param record sha3 key
+   * @param value new value is bytes32
+   */
+  function setBytes32Value(bytes32 record, bytes32 value) public onlyOwner {
+    Bytes32Values[record] = value;
+  }
+
+  /**
+   * @dev delete the value inside Bytes32Values for the given key 
+   * @param record sha3 key
+   */
+  function deleteBytes32Value(bytes32 record) public onlyOwner {
+    delete Bytes32Values[record];
+  }
+
+  /**
+   * @dev change the value inside BooleanValues for the given key 
+   * @param record sha3 key
+   * @param value new value is bool
+   */
+  function setBooleanValue(bytes32 record, bool value) public onlyOwner {
+    BooleanValues[record] = value;
+  }
+
+  /**
+   * @dev delete the value inside BooleanValues for the given key 
+   * @param record sha3 key 
+   */
+  function deleteBooleanValue(bytes32 record) public onlyOwner {
+    delete BooleanValues[record];
+  }
+}
+----
+
+*CounterContractV1*
+----
+pragma solidity ^0.4.13;
+
+import "zeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "EternalStorage.sol";
+
+contract CounterWithEternalStorage is Pausable {
+
+  EternalStorage eternalStorage;
+
+  function CounterWithEternalStorage() {
+    eternalStorage = new EternalStorage();
+  }
+
+  function getCount() public constant returns(uint256) {
+    return eternalStorage.UIntValues(keccak256("counter"));
+  } 
+    
+  function increment() public whenNotPaused returns (uint) {
+    uint count = getCount();
+    uint value = count.add(1);
+    eternalStorage.setUIntValue(keccak256("counter"), value);
+    return value;
+  }
+
+  function transferEternalStorageOwnership(address recipient) public onlyOwner whenNotPaused {
+    Ownable(eternalStorage).transferOwnership(recipient);
+  }
+}
+----
+
+So let's say for some reason you want to upgrade the *CounterContractV1* so that `increment()` increase the counter by 2; you just need to deploy a *CounterContractV2* that point to the same *EternalStorage* instead of creating it in the constructor. Then you call `transferEternalStorageOwnership(AddressOfCounterContractV2)` on *CounterContractV1* with the address of *CounterContractV2* as argument.
+
+*CounterContractV2*
+----
+pragma solidity ^0.4.13;
+
+import "zeppelin-solidity/contracts/lifecycle/Pausable.sol";
+import "EternalStorage.sol";
+
+contract CounterWithEternalStorage is Pausable {
+
+  EternalStorage eternalStorage;
+
+  function CounterWithEternalStorage(address _eternalStorage) {
```

Comment:
The variable naming is inconsistent `_eternalStorage`. This one uses `_` whereas the others don't.  I would recommend we be consistent across.

Reply:
It's to avoid collision between variables names..

Reply:
I agree it avoids a collision. Why not name all the parameters to either have `_` or avoid the names having collision.